core/transfer/src/main/java/org/eclipse/dataspaceconnector/transfer/core/commandqueue/BoundedCommandQueue.java

```diff
@@ -0,0 +1,89 @@
+package org.eclipse.dataspaceconnector.transfer.core.commandqueue;
+
+import org.eclipse.dataspaceconnector.spi.transfer.command.Command;
+import org.eclipse.dataspaceconnector.spi.transfer.command.CommandQueue;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * In-memory implementation of the {@link CommandQueue} that is backed by a bounded queue. This means that attempting
+ * to add elements to an already full queue will fail and raise an {@link IllegalStateException}.
+ * <p>
+ * This queue is threadsafe, so every operation on it can be assumed atomic.
+ */
+public class BoundedCommandQueue implements CommandQueue {
+    private final ArrayBlockingQueue<Command> queue;
```

Comment:
I would define it as a `Queue<Command>`, since it's the interface that we need

Reply:
It's fine to declare the field as `Queue`, however supplying anything other than an `ArrayBlockingQueue` might change the behaviour of the `BoundedCommandQueue`, so we need to keep that in mind.